citest/base/bindings.py

```diff
@@ -0,0 +1,376 @@
+    """
+    self.__overrides.update({_normalize_key(name): value
+                             for name, value in values.items()})
+
+  def add_config_file(self, path):
+    """Explicitly add a configuration file to the bindings.
+
+    This is a lazy call. The bindings will be read on build().
+    """
+    self.__config_files.append(path)
+    self.__dummy_config_parser.read([path]) # only to eagerly discover errors.
+
+  def __is_config_name(self, config_parser, key):
+    """Determine if a normalized key is a known configuration value.
+
+    Args:
+      config_parser: [ConfigParser] Knows about options.
+      key: [string] Normalized key.
+    """
+    if key in self.__defaults or key in self.__overrides:
+      return True
+    for section in config_parser.sections():
+      if key in config_parser.options(section):
+        return True
+    return False
+
+  def __infer_options(self, config_parser, extra):
+    """Add bindings inferred from unexpected command-line arguments.
+
+    Args:
+      config_parser: [ConfigParser] Knows about options.
+      extra: [list of string] Extra arguments from parsing command line.
+    """
+
+    result = {}
+    values = []
+    name = None
+
+    for param in extra:
+      if not param.startswith('--'):
+        values.append(param)
+        continue
+
+      # Write previous entry.
+      if name and True or self.__is_config_name(config_parser, name):
+        if not values:
+          values = ['True']
+        if len(values) == 1:
+          result[name] = values[0]
+        else:
+          result[name] = values
+
+      # Start next entry.
+      param_parts = param[2:].split('=', 1)
+      name = _normalize_key(param_parts[0])
+      values = []
+      if len(param_parts) == 2:
+        values = param_parts[1:2]
+
+    if True or self.__is_config_name(config_parser, name):
+      if not values:
+        values = ['True']
+      if len(values) == 1:
+        result[name] = values[0]
+      else:
+        result[name] = values
+
+    return result
+
+  def build(self, section=None, infer=True):
+    """Builds a new ConfigurationBindings instance."""
+    parser = argparse.ArgumentParser()
+    for arg in self.__arguments:
+      parser.add_argument(arg[0], **arg[1])
+    options = parser.parse_known_args()
+
+    config_parser = ConfigParser.RawConfigParser(defaults=self.__defaults)
+    config_parser.read(self.__config_files)
+    flags = {}
+    flags.update(vars(options[0]))
+    extra = options[1]
+    if infer and extra:
+      flags.update(self.__infer_options(config_parser, extra))
+    flags.update({key.lower(): value
+                  for key, value in self.__overrides.items()})
+
+    bindings = ConfigurationBindings(
+        config_parser, overrides=flags, section=section)
+    return bindings
+
+  def _exists(self, path):
+    """Helper function to determine if a path exists to facilitate testing."""
+    return os.path.exists(path)
```

Comment:
It's easy to intercept to test what paths are being considered.
I didnt write the test yet. Turns out there is a trick because the test doesnt always run as a module so it's behavior changes depending on how it is run so not sure how to actually write the test. I verified it manually for the time being.